File: leetcode/560.subarray-sum-equals-k.py
#  Tag: Array, Hash Table, Prefix Sum
#  Time: O(N)
#  Space: O(N)
#  Ref: -
#  Note: -

#  Given an array of integers nums and an integer k, return the total number of subarrays whose sum equals to k.
#  A subarray is a contiguous non-empty sequence of elements within an array.
#   
#  Example 1:
#  Input: nums = [1,1,1], k = 2
#  Output: 2
#  Example 2:
#  Input: nums = [1,2,3], k = 3
#  Output: 2
#  
#   
#  Constraints:
#  
#  1 <= nums.length <= 2 * 104
#  -1000 <= nums[i] <= 1000
#  -107 <= k <= 107
#  
#  

from collections import defaultdict
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        n = len(nums)
        prefix = 0
        mapping = defaultdict(int)
        mapping[0] = 1

        res = 0
        for n in nums:
            prefix += n
            res += mapping[prefix - k]
            mapping[prefix] += 1

        return res

File: leetcode/560.subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int prefix_sum = 0;
        unordered_map<int, vector<int>> map;
        map[k] = vector<int>(1, -1);
        int count = 0;
        
        for (auto i = 0; i < nums.size(); ++i)
        {
            prefix_sum += nums[i];
            if (map.count(prefix_sum) != 0)
            {
                count += map[prefix_sum].size();
            }
            map[prefix_sum + k].push_back(i);
        }
        
        return count;
    }
};